mlir/lib/Catalyst/Transforms/InlineNestedModules.cpp

```diff
@@ -0,0 +1,493 @@
+
+        bool run = _stopAfterStep >= 2 || _stopAfterStep == 0;
+        if (run &&
+            failed(applyPatternsAndFoldGreedily(symbolTable, std::move(renameFunctions), config))) {
+            signalPassFailure();
+        }
+
+        RewritePatternSet inlineNested(context);
+        inlineNested.add<InlineNestedModule>(context);
+        run = _stopAfterStep >= 3 || _stopAfterStep == 0;
+        if (run &&
+            failed(applyPatternsAndFoldGreedily(symbolTable, std::move(inlineNested), config))) {
+            signalPassFailure();
+        }
+
+        mlir::DenseMap<SymbolRefAttr, SymbolRefAttr> old_to_new;
+        for (auto &region : symbolTable->getRegions()) {
+            for (auto &block : region.getBlocks()) {
+                for (auto &op : block) {
+                    if (!isa<SymbolOpInterface>(op))
+                        continue;
+                    auto hasQualifiedName = op.hasAttr(fullyQualifiedNameAttr);
+                    if (!hasQualifiedName)
+                        continue;
+
+                    SymbolRefAttr old = op.getAttrOfType<SymbolRefAttr>(fullyQualifiedNameAttr);
+                    auto symbol = cast<SymbolOpInterface>(op);
+                    SymbolRefAttr _new = SymbolRefAttr::get(symbol);
+                    old_to_new.insert({old, _new});
+                }
+            }
+        }
+
+        RewritePatternSet nestedToFlat(context);
+        nestedToFlat.add<NestedToFlatCallPattern, SymbolReplacerPattern, ZNEReplacerPattern>(
+            context, &old_to_new);
+        run = _stopAfterStep >= 4 || _stopAfterStep == 0;
+        if (run &&
+            failed(applyPatternsAndFoldGreedily(symbolTable, std::move(nestedToFlat), config))) {
+            signalPassFailure();
+        }
+
+        RewritePatternSet cleanup(context);
+        cleanup.add<CleanupPattern>(context);
+        run = _stopAfterStep >= 5 || _stopAfterStep == 0;
+        if (run && failed(applyPatternsAndFoldGreedily(symbolTable, std::move(cleanup), config))) {
+            signalPassFailure();
+        }
+    }
+};
+
+#define GEN_PASS_DEF_INLINENESTEDMODULEPASS
+#define GEN_PASS_DECL_INLINENESTEDMODULEPASS
+#include "Catalyst/Transforms/Passes.h.inc"
+
+struct InlineNestedModulePass : impl::InlineNestedModulePassBase<InlineNestedModulePass> {
+    using InlineNestedModulePassBase::InlineNestedModulePassBase;
+
+    void runOnOperation() final
+    {
+        MLIRContext *ctx = &getContext();
+        auto op = getOperation();
+        auto pm = PassManager::on<ModuleOp>(ctx);
+        if (stopAfterStep >= 1 || stopAfterStep == 0) {
+            OpPassManager &nestedModulePM = pm.nestAny();
+            nestedModulePM.addPass(std::make_unique<AnnotateWithFullyQualifiedNamePass>());
+        }
+
+        if (stopAfterStep >= 2 || stopAfterStep == 0) {
+            pm.addPass(std::make_unique<InlineNestedSymbolTablePass>(stopAfterStep));
+        }
```

Comment:
I originally wanted to make all of this a pass using the pass manager. But then I found out that an invariant between passes is that the IR is verifies between passes. This is not true within a single pass. Mutliple patterns may be applied within a single pass and while the pass is still matching and rewriting, the IR is allowed to be in a state that won't verify. I learned later that this can be disabled by using [`PassManager::enableVerifier(false)`](https://mlir.llvm.org/doxygen/classmlir_1_1PassManager.html#af8f23f27fdf0da808e62363126abcd68). I am still thinking about what is the best way to write passes and tests. Should every pattern have a pass that we can then call from the command line to lit test? And then there are some passes which just use other passes? Or patterns?

If you would allow, I think keeping this separation right now is ok. Let me play a little bit more while I figure out the best way. I might rewrite some parts of this pass to avoid some passes and have it more just in data structures. But I also want to make some upstream contributions regarding renaming functions inside nested modules.

Reply:
> Should every pattern have a pass that we can then call from the command line to lit test? And then there are some passes which just use other passes? Or patterns?

That's a good question and I don't have enough experience to know what is considered good practice here. One thing I will say is that "one pattern - one pass" does not make sense to me. Having those be separate abstractions becomes somewhat meaningless then. I wonder if someone in the MLIR discord could answer this question, especially with regards to testing. If you would like to test patterns individually, but you don't have access to them from opt directly, how do you do so? 

What we can do is search for `applyPatternsAndFoldGreedily` in upstream MLIR, and from a quick look you will always find several or even a large number of pattern. This makes sense to me as well as otherwise the entire notion pattern rewriting and folding iteratively until a fixed point makes a lot less sense.

The other questions we could ask is wether we should even use patterns to begin. They make a lot of sense for things that need to be applied iteratively, but say we want to apply a very specific transformation once, does it even make sense to use a pattern?

Reply:
I already asked this question a while back, but it was somewhat unsatisfactory. Because I am interested in lit testing patterns that may override each other. In that sense one pattern == one pass, make sense to lit test. Otherwise, I need the command line option stating at which step of the pass to stop.

The answer I received was that essentially you can test multiple patterns at once easily is they do not override each other. This I agree with.

Reply:
Seems a bit broken to have to go against intended abstraction patterns for testing purposes 🤔

Reply:
> The other questions we could ask is wether we should even use patterns to begin. They make a lot of sense for things that need to be applied iteratively, but say we want to apply a very specific transformation once, does it even make sense to use a pattern?

I think you are asking more about the pattern driver, not the pattern itself. But I agree with the idea! [I think that the greedy rewriter already has enough configuration to do exactly what you want.](https://mlir.llvm.org/doxygen/classmlir_1_1GreedyRewriteConfig.html) But I am still unsure, it seems some of the jargon does not match what I would expect. (An iteration of the worklist seems a bit of a weird unit to count, because I expect the worklist to only insert elements back into it if there was a change in an element).

We could write our own pattern driver to narrow down where in which specific ops it should be applied. I thought this already existed in MLIR but it must have been deleted or I imagined it :sweat_smile:

Reply:
Good discussion! For this PR, I don't think we need to change much. You already have the system in place to pass a config option to stop at a certain stage, so we can just go with that :)
The only thing I find odd is that the first step is run in its own pass, and the rest aren't. So personally I'd find it cleaner to use the same method throughout, like not using nested PMs at all, but if you have a good reason to have this split then we can go with that.

Reply:
> We could write our own pattern driver to narrow down where in which specific ops it should be applied. I thought this already existed in MLIR but it must have been deleted or I imagined it

[Found it:](https://mlir.llvm.org/doxygen/GreedyPatternRewriteDriver_8h_source.html)

```
/// Rewrite the specified ops by repeatedly applying the highest benefit
 /// patterns in a greedy worklist driven manner until a fixpoint is reached.
 ///
 /// The greedy rewrite may prematurely stop after a maximum number of
 /// iterations, which can be configured in the configuration parameter.
 ///
 /// Also performs folding and simple dead-code elimination before attempting to
 /// match any of the provided patterns.
 ///
 /// Newly created ops and other pre-existing ops that use results of rewritten
 /// ops or supply operands to such ops are also processed, unless such ops are
 /// excluded via `config.strictMode`. Any other ops remain unmodified (i.e.,
 /// regardless of `strictMode`).
 ///
 /// In addition to strictness, a region scope can be specified. Only ops within
 /// the scope are simplified. This is similar to `applyPatternsAndFoldGreedily`,
 /// where only ops within the given region/op are simplified by default. If no
 /// scope is specified, it is assumed to be the first common enclosing region of
 /// the given ops.
 ///
 /// Note that ops in `ops` could be erased as result of folding, becoming dead,
 /// or via pattern rewrites. If more far reaching simplification is desired,
 /// `applyPatternsAndFoldGreedily` should be used.
 ///
 /// Returns "success" if the iterative process converged (i.e., fixpoint was
 /// reached) and no more patterns can be matched. `changed` is set to "true" if
 /// the IR was modified at all. `allOpsErased` is set to "true" if all ops in
 /// `ops` were erased.
 LogicalResult
 [applyOpPatternsAndFold](https://mlir.llvm.org/doxygen/namespacemlir.html#a10b86ede77ff793a77a2f26ac8755313)(ArrayRef<Operation *> ops,
                        const FrozenRewritePatternSet &patterns,
                        GreedyRewriteConfig config = GreedyRewriteConfig(),
                        bool *changed = nullptr, bool *allErased = nullptr);
```

Reply:
Interesting! What's your conclusion with regards to the way the steps are split between PMs and patterns?

Reply:
I think it is ok to keep the current structure. It reflects that the passes are specific for nested symboltables. Otherwise, I would need to do some filtering by hand, which is what the pass is already doing.